In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sbn

df = pd.read_csv(r"C:\Users\riya sachdev\Desktop\Spotify EDA Project\spotify_data clean.csv")
df


track_id                                       track_name  \
0     3EJS5LyekDim1Tf5rBFmZl                    Trippy Mane (ft. Project Pat)   
1     1oQW6G2ZiwMuHqlPpP27DB                                             OMG!   
2     7mdkjzoIYlf1rx9EtBpGmU                                      Hard 2 Find   
3     67rW0Zl7oB3qEpD5YWWE5w  Still Get Like That (ft. Project Pat & Starrah)   
4     15xptTfRBrjsppW0INUZjf                            ride me like a harley   
...                      ...                                              ...   
8577  0AQquaENerGps8BQmbPw14                                         Big Iron   
8578  4f8hBeMXMvssn6HtFAtblo                                          El Paso   
8579  0Vy7wsXNFrbNc6UTWoScnM                                 Over the Rainbow   
8580  760clbeDBWmBsBLbszWuNZ                                        I'm A Man   
8581  61GEP8lryEfcuEgBMbRmNi                               Alone And Forsaken   

      track_number  track_popularity  explicit    artist_name  \
0                4                 0      True          Diplo   
1                1                 0      True       Yelawolf   
2                1                 4      True      Riff Raff   
3                8                30      True          Diplo   
4                2                 0      True        Rumelis   
...            ...               ...       ...            ...   
8577             1                71     False  Marty Robbins   
8578            10                64     False  Marty Robbins   
8579             5                 8     False    The Mystics   
8580             2                55     False     Bo Diddley   
8581            11                45     False  Hank Williams   

      artist_popularity  artist_followers  \
0                    77           2812821   
1                    64           2363438   
2                    48            193302   
3                    77           2813710   
4                    48              8682   
...                 ...               ...   
8577                 60            626733   
8578                 61            626733   
8579                 27             18184   
8580                 44            333376   
8581                 55           1067339   

                                          artist_genres  \
0                                            moombahton   
1                     country hip hop, southern hip hop   
2                                                   NaN   
3                                            moombahton   
4                                              dark r&b   
...                                                 ...   
8577                    classic country, outlaw country   
8578                    classic country, outlaw country   
8579                                            doo-wop   
8580    blues, classic blues, rock and roll, rockabilly   
8581  classic country, honky tonk, traditional count...   

                    album_id                            album_name  \
0     5QRFnGnBeMGePBKF2xTz5z                    d00mscrvll, Vol. 1   
1     4SUmmwnv0xTjRcLdjczGg2                                  OMG!   
2     3E3zEAL8gUYWaLYB9L7gbp                           Hard 2 Find   
3     5QRFnGnBeMGePBKF2xTz5z                    d00mscrvll, Vol. 1   
4     06FDIpSHYmZAZoyuYtc7kd   come closer / ride me like a harley   
...                      ...                                   ...   
8577  3kQpBS26lAj0A0VGl1snRl    Gunfighter Ballads And Trail Songs   
8578  3kQpBS26lAj0A0VGl1snRl    Gunfighter Ballads And Trail Songs   
8579  2ifB9Xjp9DdpqLlYlY60QW                Presenting The Mystics   
8580  1cbtDEwxCjMhglb49OgNBR                            Bo Diddley   
8581  3cusZESjkIDnDXyQwbpSsT  Moanin' The Blues (Expanded Edition)   

     album_release_date  album_total_tracks   album_type  track_duration_min  
0            2025-10-31                   9        album                1.55  
1            2025

In [2]:
print("Number of columns:",len(df.columns))
print("And their names are:", df.columns)

Number of columns: 15
And their names are: Index(['track_id', 'track_name', 'track_number', 'track_popularity',
       'explicit', 'artist_name', 'artist_popularity', 'artist_followers',
       'artist_genres', 'album_id', 'album_name', 'album_release_date',
       'album_total_tracks', 'album_type', 'track_duration_min'],
      dtype='object')


In [3]:
df.dtypes

track_id               object
track_name             object
track_number            int64
track_popularity        int64
explicit                 bool
artist_name            object
artist_popularity       int64
artist_followers        int64
artist_genres          object
album_id               object
album_name             object
album_release_date     object
album_total_tracks      int64
album_type             object
track_duration_min    float64
dtype: object

## while examine data types of fields we realized that album_release_date is currently an object and we have to convert it to date_time for time based analysis

In [5]:
df["album_release_date"] = pd.to_datetime(df["album_release_date"],errors = "coerce")
df.dtypes

track_id                      object
track_name                    object
track_number                   int64
track_popularity               int64
explicit                        bool
artist_name                   object
artist_popularity              int64
artist_followers               int64
artist_genres                 object
album_id                      object
album_name                    object
album_release_date    datetime64[ns]
album_total_tracks             int64
album_type                    object
track_duration_min           float64
dtype: object

## Null Handling

In [6]:
print("Total Null Values in our Data Set :", df.isnull().sum().sum())
print("Total Null Values in our Data Set (column wise):\n\n")
print(df.isnull().sum())

Total Null Values in our Data Set : 11946
Total Null Values in our Data Set (column wise):


track_id                 0
track_name               0
track_number             0
track_popularity         0
explicit                 0
artist_name              3
artist_popularity        0
artist_followers         0
artist_genres         3361
album_id                 0
album_name               0
album_release_date    8582
album_total_tracks       0
album_type               0
track_duration_min       0
dtype: int64


In [8]:
df = df.dropna(subset=["artist_name"])

In [9]:
df["artist_genres"] = df["artist_genres"].fillna("Unknown")

In [11]:
print("Total Null Values in our Data Set after fillna :", df.isnull().sum())

Total Null Values in our Data Set after fillna : track_id                 0
track_name               0
track_number             0
track_popularity         0
explicit                 0
artist_name              0
artist_popularity        0
artist_followers         0
artist_genres            0
album_id                 0
album_name               0
album_release_date    8579
album_total_tracks       0
album_type               0
track_duration_min       0
dtype: int64


#### Insights
All columns are complete after filling missing values, except **album_release_date**, which has **8,579 missing entries**. This is perfectly fine to have null values **(NaT - Not a Time)** in this column, as some albums may not have a **recorded release date**.

In [13]:
df

track_id                                       track_name  \
0     3EJS5LyekDim1Tf5rBFmZl                    Trippy Mane (ft. Project Pat)   
1     1oQW6G2ZiwMuHqlPpP27DB                                             OMG!   
2     7mdkjzoIYlf1rx9EtBpGmU                                      Hard 2 Find   
3     67rW0Zl7oB3qEpD5YWWE5w  Still Get Like That (ft. Project Pat & Starrah)   
4     15xptTfRBrjsppW0INUZjf                            ride me like a harley   
...                      ...                                              ...   
8577  0AQquaENerGps8BQmbPw14                                         Big Iron   
8578  4f8hBeMXMvssn6HtFAtblo                                          El Paso   
8579  0Vy7wsXNFrbNc6UTWoScnM                                 Over the Rainbow   
8580  760clbeDBWmBsBLbszWuNZ                                        I'm A Man   
8581  61GEP8lryEfcuEgBMbRmNi                               Alone And Forsaken   

      track_number  track_popularity  explicit    artist_name  \
0                4                 0      True          Diplo   
1                1                 0      True       Yelawolf   
2                1                 4      True      Riff Raff   
3                8                30      True          Diplo   
4                2                 0      True        Rumelis   
...            ...               ...       ...            ...   
8577             1                71     False  Marty Robbins   
8578            10                64     False  Marty Robbins   
8579             5                 8     False    The Mystics   
8580             2                55     False     Bo Diddley   
8581            11                45     False  Hank Williams   

      artist_popularity  artist_followers  \
0                    77           2812821   
1                    64           2363438   
2                    48            193302   
3                    77           2813710   
4                    48              8682   
...                 ...               ...   
8577                 60            626733   
8578                 61            626733   
8579                 27             18184   
8580                 44            333376   
8581                 55           1067339   

                                          artist_genres  \
0                                            moombahton   
1                     country hip hop, southern hip hop   
2                                               Unknown   
3                                            moombahton   
4                                              dark r&b   
...                                                 ...   
8577                    classic country, outlaw country   
8578                    classic country, outlaw country   
8579                                            doo-wop   
8580    blues, classic blues, rock and roll, rockabilly   
8581  classic country, honky tonk, traditional count...   

                    album_id                            album_name  \
0     5QRFnGnBeMGePBKF2xTz5z                    d00mscrvll, Vol. 1   
1     4SUmmwnv0xTjRcLdjczGg2                                  OMG!   
2     3E3zEAL8gUYWaLYB9L7gbp                           Hard 2 Find   
3     5QRFnGnBeMGePBKF2xTz5z                    d00mscrvll, Vol. 1   
4     06FDIpSHYmZAZoyuYtc7kd   come closer / ride me like a harley   
...                      ...                                   ...   
8577  3kQpBS26lAj0A0VGl1snRl    Gunfighter Ballads And Trail Songs   
8578  3kQpBS26lAj0A0VGl1snRl    Gunfighter Ballads And Trail Songs   
8579  2ifB9Xjp9DdpqLlYlY60QW                Presenting The Mystics   
8580  1cbtDEwxCjMhglb49OgNBR                            Bo Diddley   
8581  3cusZESjkIDnDXyQwbpSsT  Moanin' The Blues (Expanded Edition)   

     album_release_date  album_total_tracks   album_type  track_duration_min  
0                   NaT                   9        album                1.55  
1                

### Final Objectives of Spotify Data Analysis 
* To understand the structure, size, and quality of the dataset using basic inspection techniques such as **info(), shape, and describe()**.
* To analyze the statistical distribution of numerical variables including **track popularity, track duration, artist popularity, artist followers, and album total tracks**.
* To visualize the distribution of key variables using **histograms and density plots** to identify trends, skewness, and outliers.
* To examine the proportion of explicit and non-explicit tracks using **pie charts** and evaluate their impact on track popularity.
* To explore relationships between numerical features using scatter plots, such as:

  **Track Duration versus Track popularity**

  **Artist follower versus artist popularity**
  
* To compare track popularity across different album types using **box plots and categorical visualizations**.
* To analyze music release trends over time by extracting release years from album release dates and visualizing changes in music production patterns.
* To identify correlations among numerical variables using **correlation matrices and heatmaps** in order to understand interdependencies within the dataset.
* To identify the Top 5 artists based on different performance criteria, including:
  
  **Average track popularity (consistency)**
  
  **Maximum track popularity (viral hits)**
  
  **Total number of tracks (productivity)**
  
  **Artist popularity score (platform-level influence)**
  
  **Number of artist followers (audience reach)**

* To perform a combined artist ranking analysis by integrating multiple metrics such as track popularity, artist popularity, and followers to identify overall top-performing artists.
* To extract meaningful insights and actionable conclusions that can help artists, producers, and music streaming platforms improve content strategy, audience engagement, and release planning.h)
ty
te planning.

